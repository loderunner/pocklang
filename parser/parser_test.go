package parser

import (
	"strings"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
)

func TestParser(t *testing.T) {
	type testCase struct {
		input    string
		expected Expr
	}
	cases := []testCase{
		{
			input: "1 == 2",
			expected: BinaryExpr{
				Op:    Eq,
				Left:  LiteralExpr{Token: Token{Type: Integer, Lexeme: "1", IntegerValue: 1}},
				Right: LiteralExpr{Token: Token{Type: Integer, Lexeme: "2", IntegerValue: 2}},
			},
		},
	}

	t.Parallel()
	for _, c := range cases {
		t.Run(c.input, func(t *testing.T) {
			tokens, err := Scan(strings.NewReader(c.input))
			assert.NoErrorf(t, err, "")
			expr, err := Parse(tokens)
			assert.NoErrorf(t, err, "")
			assert.Equal(t, c.expected, expr)
		})
	}
}

func TestParserSnapshots(t *testing.T) {
	cases := []string{
		"hello.world > 3",
		`"hello" != "world"`,
		"((3+2) - 14) == -19",
		`123.45 * "d" < asdrg`,
		"true && false || null == (42 / 2)",
	}
	t.Parallel()
	for _, c := range cases {
		t.Run(c, func(t *testing.T) {
			tokens, err := Scan(strings.NewReader(c))
			assert.NoErrorf(t, err, "")
			expr, err := Parse(tokens)
			assert.NoErrorf(t, err, "")
			snaps.MatchSnapshot(t, expr)
		})
	}
}

func TestParserErrors(t *testing.T) {
	cases := []string{
		"",
		"hello.",
		".hello",
		"12 <",
		"12.hello",
		"4 << 54",
		"(41 + d",
		`(""+)`,
		"--3",
		"3*",
		"true && || false",
		"true || && false",
	}
	t.Parallel()
	for _, c := range cases {
		t.Run(c, func(t *testing.T) {
			tokens, err := Scan(strings.NewReader(c))
			assert.NoErrorf(t, err, "")
			_, err = Parse(tokens)
			assert.Error(t, err)
			snaps.MatchSnapshot(t, err)
		})
	}
}
