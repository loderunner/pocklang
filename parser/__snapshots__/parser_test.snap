
[TestParserSnapshots/hello.world_>_3 - 1]
parser.BinaryExpr{
    Op:   Gt,
    Left: parser.GetExpr{
        Names: {"hello", "world"},
    },
    Right: parser.LiteralExpr{
        Token: parser.Token{Type:Integer, Lexeme:"3", IntegerValue:3, DecimalValue:0, StringValue:"", IdentifierValue:""},
    },
}
---

[TestParserSnapshots/"hello"_!=_"world" - 1]
parser.BinaryExpr{
    Op:   Neq,
    Left: parser.LiteralExpr{
        Token: parser.Token{Type:String, Lexeme:"\"hello\"", IntegerValue:0, DecimalValue:0, StringValue:"hello", IdentifierValue:""},
    },
    Right: parser.LiteralExpr{
        Token: parser.Token{Type:String, Lexeme:"\"world\"", IntegerValue:0, DecimalValue:0, StringValue:"world", IdentifierValue:""},
    },
}
---

[TestParserSnapshots/((3+2)_-_14)_==_-19 - 1]
parser.BinaryExpr{
    Op:   Eq,
    Left: parser.GroupExpr{
        Expr: parser.BinaryExpr{
            Op:   Minus,
            Left: parser.GroupExpr{
                Expr: parser.BinaryExpr{
                    Op:   Plus,
                    Left: parser.LiteralExpr{
                        Token: parser.Token{Type:Integer, Lexeme:"3", IntegerValue:3, DecimalValue:0, StringValue:"", IdentifierValue:""},
                    },
                    Right: parser.LiteralExpr{
                        Token: parser.Token{Type:Integer, Lexeme:"2", IntegerValue:2, DecimalValue:0, StringValue:"", IdentifierValue:""},
                    },
                },
            },
            Right: parser.LiteralExpr{
                Token: parser.Token{Type:Integer, Lexeme:"14", IntegerValue:14, DecimalValue:0, StringValue:"", IdentifierValue:""},
            },
        },
    },
    Right: parser.UnaryExpr{
        Op:   Minus,
        Expr: parser.LiteralExpr{
            Token: parser.Token{Type:Integer, Lexeme:"19", IntegerValue:19, DecimalValue:0, StringValue:"", IdentifierValue:""},
        },
    },
}
---

[TestParserSnapshots/123.45_*_"d"_<_asdrg - 1]
parser.BinaryExpr{
    Op:   Lt,
    Left: parser.BinaryExpr{
        Op:   Star,
        Left: parser.LiteralExpr{
            Token: parser.Token{Type:Decimal, Lexeme:"123.45", IntegerValue:0, DecimalValue:123.45, StringValue:"", IdentifierValue:""},
        },
        Right: parser.LiteralExpr{
            Token: parser.Token{Type:String, Lexeme:"\"d\"", IntegerValue:0, DecimalValue:0, StringValue:"d", IdentifierValue:""},
        },
    },
    Right: parser.GetExpr{
        Names: {"asdrg"},
    },
}
---

[TestParserSnapshots/true_&&_false_||_null_==_(42_/_2) - 1]
parser.BinaryExpr{
    Op:   Null,
    Left: parser.BinaryExpr{
        Op:   False,
        Left: parser.LiteralExpr{
            Token: parser.Token{Type:True, Lexeme:"true", IntegerValue:0, DecimalValue:0, StringValue:"", IdentifierValue:""},
        },
        Right: parser.LiteralExpr{
            Token: parser.Token{Type:False, Lexeme:"false", IntegerValue:0, DecimalValue:0, StringValue:"", IdentifierValue:""},
        },
    },
    Right: parser.BinaryExpr{
        Op:   Eq,
        Left: parser.LiteralExpr{
            Token: parser.Token{Type:Null, Lexeme:"null", IntegerValue:0, DecimalValue:0, StringValue:"", IdentifierValue:""},
        },
        Right: parser.GroupExpr{
            Expr: parser.BinaryExpr{
                Op:   Slash,
                Left: parser.LiteralExpr{
                    Token: parser.Token{Type:Integer, Lexeme:"42", IntegerValue:42, DecimalValue:0, StringValue:"", IdentifierValue:""},
                },
                Right: parser.LiteralExpr{
                    Token: parser.Token{Type:Integer, Lexeme:"2", IntegerValue:2, DecimalValue:0, StringValue:"", IdentifierValue:""},
                },
            },
        },
    },
}
---

[TestParserErrors/#00 - 1]
&fmt.wrapError{
    msg: "parse error: unexpected end of expression",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/hello. - 1]
&fmt.wrapError{
    msg: "parse error at '': expected identifier after '.'",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/.hello - 1]
&fmt.wrapError{
    msg: "parse error at '.': unexpected token",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/12_< - 1]
&fmt.wrapError{
    msg: "parse error: unexpected end of expression",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/12.hello - 1]
&fmt.wrapError{
    msg: "parse error at 'hello': expected end of expression",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/4_<<_54 - 1]
&fmt.wrapError{
    msg: "parse error at '<': unexpected token",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/(41_+_d - 1]
&fmt.wrapError{
    msg: "parse error: missing closing parenthesis",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/(""+) - 1]
&fmt.wrapError{
    msg: "parse error at ')': unexpected token",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/--3 - 1]
&fmt.wrapError{
    msg: "parse error at '-': unexpected token",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/3* - 1]
&fmt.wrapError{
    msg: "parse error: unexpected end of expression",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/true_&&_||_false - 1]
&fmt.wrapError{
    msg: "parse error at '||': unexpected token",
    err: &errors.errorString{s:"parse error"},
}
---

[TestParserErrors/true_||_&&_false - 1]
&fmt.wrapError{
    msg: "parse error at '&&': unexpected token",
    err: &errors.errorString{s:"parse error"},
}
---
